File: java/leetcode/dp/target_sum.java
/**
 * 494. Target Sum
 * ou are given a list of non-negative integers, a1, a2, ..., an, and a target, S. Now you have 2 symbols + and -. For each integer, you should choose one from + and - as its new symbol.

Find out how many ways to assign symbols to make sum of integers equal to target S.

Example 1:

Input: nums is [1, 1, 1, 1, 1], S is 3. 
Output: 5
Explanation: 

-1+1+1+1+1 = 3
+1-1+1+1+1 = 3
+1+1-1+1+1 = 3
+1+1+1-1+1 = 3
+1+1+1+1-1 = 3

There are 5 ways to assign symbols to make the sum of nums be target 3.
Note:

The length of the given array is positive and will not exceed 20.
The sum of elements in the given array will not exceed 1000.
Your output answer is guaranteed to be fitted in a 32-bit integer.
 */

public class target_sum {

    // sum(P) - sum(N) = target
    // sum(P) + sum(N) + sum(P) - sum(N) = target + sum(P) + sum(N)
    // 2 * sum(P) = target + sum(nums)
    
    public int findTargetSumWays(int[] nums, int S) {
        int sum = computeArraySum(nums);
        if (sum < S || (sum+S) % 2 == 1) {
            return 0;
        }
        int W = (sum + S) / 2;
        int[] dp = new int[W+1];
        dp[0] = 1;
        for (int num : nums) {
            for (int i = W; i >= num; i--) {
                dp[i] = dp[i] + dp[i-num];
            }
        }
        return dp[W];
    }

    private int computeArraySum(int[] nums) {
        int sum = 0;
        for (int num : nums) {
            sum += num;
        }
        return sum;
    }

    // Solution 2 DFS
    public int findTargetSumWays2(int[] nums, int S) {
        return findTargetSumWays(nums, 0, S);
    }

    private int findTargetSumWays(int[] nums, int start, int S) {
        if (start == nums.length) {
            return S == 0 ? 1 : 0;
        }
        return findTargetSumWays(nums, start+1, S+nums[start]) + findTargetSumWays(nums, start+1, S-nums[start]);
    }

}

File: golang/leetcode/dp/target_sum.go
package dp

// 494. Target Sum
// P +, N -
// sum(P) - sum(N) = target
// sum(P) - sum(N) + sum(N) + sum(P) = target + sum(P) + sum(N)
// 2 * sum(P) = target + sum(nums)

func findTargetSumWays(nums []int, S int) int {
	sum := computeArraySum(nums)
	if sum < S || (sum+S)%2 == 1 {
		return 0
	}
	W := (sum + S) / 2
	dp := make([]int, W+1)
	dp[0] = 1
	for _, num := range nums {
		for i := W; i >= num; i-- {
			dp[i] = dp[i] + dp[i-num]
		}
	}
	return dp[W]
}

/*
func computeArraySum(nums []int) int {
	sum := 0
	for _, num := range nums {
		sum += num
	}
	return sum
}
*/
